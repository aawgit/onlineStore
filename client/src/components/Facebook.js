import React, { Component } from 'react';
import axios from 'axios';
import { Redirect } from 'react-router-dom';
import FacebookLogin from 'react-facebook-login';
import Context from '../Context';
import { API_PATH_FACEBOOK } from '../constants';

class Facebook extends Component {
	constructor(props) {
		super(props);
		this.handleResponse = this.handleResponse.bind(this);
		this.state = {
			redirect: false,
		};
	}

	static contextType = Context;

	/**
	 * Callback method of the external Facebook component
	 * Getting fired when user closes the login prompt.
	 * See more: https://github.com/keppelen/react-facebook-login	 *
	 * @param {object} response Data sent back by the Facebook API
	 */
	handleResponse(response) {
		return axios
			.post(API_PATH_FACEBOOK, {
				access_token: response.accessToken,
			})
			.then((res) => {
				this.context.setUser(res.data);
				this.setState({ redirect: '/home' });
			})
			.catch((err) => {
				this.context.setError(err);
			});
	}
	/**
	 * There is no need to test the render method itself because
	 * both components has it's own unit tests and the internal state
	 * is check elsewhere for this statement
	 */
	/* istanbul ignore next */
	render() {
		if (this.state.redirect) {
			return <Redirect to={this.state.redirect} />;
		} else {
			return (
				<FacebookLogin
					appId='1343346045860132'
					autoLoad={false}
					fields='name,email,picture'
					onClick={() => {}}
					callback={this.responseFacebook}
				/>
			);
		}
	}
}

export default Facebook;
