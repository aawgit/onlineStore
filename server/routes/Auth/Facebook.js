import dotenv from 'dotenv';
if (process.env.NODE_ENV !== 'production') dotenv.config();

import express from 'express';
import bodyParser from 'body-parser';
import jwt from 'jsonwebtoken';
import FacebookTokenStragegy from 'passport-facebook-token';
import passport from 'passport';
import User from '../../models/User';

/**
 * Logs in a user if it has an entry in the database or creates a new,
 * bases on the Facebook data recieved. Sends back the user object
 * from the database with `JWT token` and given `_id`. This allows
 * perform actions which require specific privileges.
 *
 * **Request must be `HTTP-POST @ <root>/api/auth/facebook`**
 *
 * @implements User Schema
 * @returns User Object
 */
const Facebook = express.Router();

Facebook.use(bodyParser.urlencoded({ extended: false }));
Facebook.use(bodyParser.json());

passport.serializeUser(function (user, cb) {
	cb(null, user);
});

passport.deserializeUser(function (obj, cb) {
	cb(null, obj);
});

passport.use(
	'facebookToken',
	new FacebookTokenStragegy(
		{
			clientID: process.env.FACEBOOK_CLIENT_ID,
			clientSecret: process.env.FACEBOOK_CLIENT_SECRET,
		},
		(accessToken, refreshToken, profile, done) => {
			return done(null, {
				name: profile._json.name,
				email: profile._json.email,
			});
		}
	)
);

Facebook.use(passport.initialize());
Facebook.use(passport.session());

Facebook.post('*', passport.authenticate('facebookToken'), (req, res) => {
	User.findOne({ email: req.user.email }, (err, user) => {
		if (err) return res.status(404).send(err);
		if (user) {
			return res.send({
				id: user._id,
				name: user.name,
				auth: true,
				token: jwt.sign({ id: user._id }, process.env.JWT_SECRET, {
					expiresIn: 86400,
				}),
			});
		} else {
			User.create(req.user, (err, user) => {
				if (err) return res.status(500).send(err);
				if (user) {
					return res.status(201).send({
						id: user._id,
						name: user.name,
						auth: true,
						token: jwt.sign({ id: user._id }, process.env.JWT_SECRET, {
							expiresIn: 86400,
						}),
					});
				}
			});
		}
	});
});

export default Facebook;
