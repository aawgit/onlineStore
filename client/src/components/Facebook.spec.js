import React from 'react';
import { shallow } from 'enzyme';
import Facebook, { API_PATH } from './Facebook';
import FacebookLogin from 'react-facebook-login';
import axios from 'axios';

jest.mock('axios');

global.console = {
	log: jest.fn(),
};

const restoreLogin = Facebook.prototype.onLoginSuccess;
const restoreResponse = Facebook.prototype.responseFacebook;

describe('<Facebook />', () => {
	afterEach(() => {
		Facebook.prototype.onLoginSuccess = restoreLogin;
		Facebook.prototype.responseFacebook = restoreResponse;
	});

	it('should render', () => {
		const wrapper = shallow(<Facebook />);
		expect(wrapper).toMatchSnapshot();
	});

	describe('responseFacebook', () => {
		let mock_onLoginSuccess, mock_response, wrapper;

		beforeEach(() => {
			mock_response = {
				accessToken: 'token1',
				name: 'name1',
				email: 'eamil',
				picture: {
					data: {
						url: 'url1',
					},
				},
			};
			mock_onLoginSuccess = jest.fn();
			Facebook.prototype.onLoginSuccess = mock_onLoginSuccess;
			wrapper = shallow(<Facebook />);
			wrapper.setState({
				redirect: false,
				user: {
					isLoggedIn: false,
					name: '',
					email: '',
					picture: '',
					accessToken: '',
				},
			});
		});

		afterEach(() => {
			mock_onLoginSuccess.mockRestore();
		});

		it('should set state and call onLoginSuccess', () => {
			wrapper.instance().responseFacebook(mock_response);
			expect(wrapper.state().user.isLoggedIn).toBeTruthy();
			expect(wrapper.state().user.accessToken).toBeTruthy();
			expect(mock_onLoginSuccess).toBeCalledTimes(1);
		});

		it('should log error if no token', () => {
			wrapper.instance().responseFacebook({ mock: { accessToken: false } });
			expect(wrapper.state().user.isLoggedIn).toBeFalsy();
			expect(wrapper.state().user.accessToken).toBeFalsy();
			expect(mock_onLoginSuccess).toBeCalledTimes(0);
		});
	});

	describe('onLoginSuccess', () => {
		let wrapper, resolve, reject, mock_user;
		beforeEach(() => {
			mock_user = { accessToken: 'accesstoken1' };
			resolve = Promise.resolve({
				data: {
					token: 'token1',
					userId: 'userid1',
					name: 'name1',
				},
			});
			reject = Promise.reject('error1');
			sessionStorage.setItem('user', null);
			wrapper = shallow(<Facebook />);
		});

		it('should resolve', async () => {
			wrapper.setState({ user: mock_user });
			axios.post.mockImplementationOnce(() => resolve);
			wrapper.instance().onLoginSuccess();
			await resolve;
			expect(axios.post).toHaveBeenCalledWith(API_PATH, {
				access_token: mock_user.accessToken,
			});
			expect(sessionStorage.getItem('user')).toBeTruthy();
			expect(wrapper.state().redirect).toBeTruthy();
			expect(wrapper.find('Redirect')).toBeTruthy();
		});

		it('should reject', async () => {
			try {
				axios.post.mockImplementationOnce(() => reject);
				await wrapper.instance().onLoginSuccess();
			} catch (e) {
				expect(global.console.log).toHaveBeenCalledWith('error1');
			}
		});
	});

	describe('<FacebookLogin />', () => {
		it('should fire callback', () => {
			const mock_responseFacebook = jest.fn();
			Facebook.prototype.responseFacebook = mock_responseFacebook;
			const wrapper = shallow(<Facebook />);
			wrapper.find(FacebookLogin).props().callback();
			expect(mock_responseFacebook).toHaveBeenCalledTimes(1);
		});
	});
});
