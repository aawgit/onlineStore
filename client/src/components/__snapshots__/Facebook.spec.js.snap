// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Facebook /> should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Facebook />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "appId": "1343346045860132",
      "autoLoad": false,
      "callback": [Function],
      "cssClass": "kep-login-facebook",
      "fields": "name,email,picture",
      "onClick": [Function],
      "size": "metro",
      "tag": "button",
      "textButton": "Login with Facebook",
      "typeButton": "button",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "appId": "1343346045860132",
        "autoLoad": false,
        "callback": [Function],
        "cssClass": "kep-login-facebook",
        "fields": "name,email,picture",
        "onClick": [Function],
        "size": "metro",
        "tag": "button",
        "textButton": "Login with Facebook",
        "typeButton": "button",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
